### DISCUSSION TITLE
how to achieve more granular polling for TI-RTOS?

### DISCUSSION THREAD

**Discussion by keffkeff** [cesanta/mongoose#1677] [Ask questions, get answers] [answered]
I noticed that when using mongoose with TI-RTOS, `mg_mgr_poll()` does not work with timeout values more granular than seconds. I later realized that in the case of TI-RTOS, `mg_millis()` becomes `time(NULL)`, which returns the UNIX time in seconds and not in milliseconds.

So my question is if it's possible to get more precise timeout values?

**Comment by keffkeff** [comment]
OK so I was able to get a more precise timeout value by using `Clock_getTicks()` from `<ti/sysbios/knl/Clock.h>` instead of `time(NULL) * 1000`. I know that `Clock_getTicks()` does not return the UNIX time, so if some function depends on the UNIX time being returned from `mg_millis()`, then that's obviously a problem. And now I see that the SNTP and DNS implementations use it for their packets, so that's probably not good news for my workaround.

Anyway, any advice would be much appreciated.

**Reply by scaprile** [reply]
#1814

**Comment by scaprile** [comment, marked as answer]
I'm checking the example code, and as far as I can see that does not use TI RTOS.
The fastest approach is perhaps you defining `MG_ENABLE_CUSTOM_MILLIS` and providing your own `uint64_t mg_millis(void)` implementation returning up time in milliseconds; as I think you already did.
It should be fine if you just return up time, ports to some architectures do just that:
https://github.com/cesanta/mongoose/blob/01390fe5406a336821ad3b02c76e1b048f83d295/src/util.c#L90-L128

**Comment by cpq** [comment]
I second Sergio's suggestion. Mongoose's software timer implementation uses `mg_millis()`, which returns a `uint64_t` value - a system uptime in milliseconds. For those environments where we can't guess a "real" monotonic clock, we just fall back to C `time()`. Implementing your own `mg_millis()` is the way to get timers with a millisecond granularity.
